### PR TITLE
[Internal] Query: Adds CosmosUndefined to represent undefined values

### PR DESCRIPTION
**Pull request by neildsh**
## Description

Currently `undefined` values are sometimes represented by `null` in the query pipeline. At other times, `undefined` values are unhandled. In this change we use a more formal approach of representing `undefined` by a strong type called `CosmosUndefined` which is a `CosmosElement`.

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Closing issues

To automatically close an issue: closes #IssueNumber

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/Distinct/DistinctHash.cs [hunk 1]
```diff
@@ -41,6 +41,7 @@ private sealed class CosmosElementHasher : ICosmosElementVisitor<UInt128, UInt12
 
             private static class HashSeeds
             {
+                public static readonly UInt128 Undefined = UInt128.Create(0xDEAD, 0xBEEF);
```
Comment (neildsh) [inline Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/Distinct/DistinctHash.cs:44]: FIXME: Need to add the correct seeds

File: Microsoft.Azure.Cosmos/src/CosmosElements/CosmosUndefined.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+#if INTERNAL
+#pragma warning disable CS1591 // Missing XML comment for publicly visible type or member
+#pragma warning disable SA1600 // Elements should be documented
+    public
+#else
+    internal
+#endif
+    sealed class CosmosUndefined : CosmosElement, IEquatable<CosmosUndefined>, IComparable<CosmosUndefined>
+    {
+        public static readonly CosmosUndefined Instance = new CosmosUndefined();
```
Comment (neildsh) [inline Microsoft.Azure.Cosmos/src/CosmosElements/CosmosUndefined.cs:21]: Make this private

File: Microsoft.Azure.Cosmos/src/Json/JsonSerializer.cs [hunk 3]
```diff
@@ -301,6 +301,16 @@ public TryCatch<object> Visit(CosmosNull cosmosNull, Type type)
                 return TryCatch<object>.FromResult(default);
             }
 
+            public TryCatch<object> Visit(CosmosUndefined cosmosUndefined, Type type)
+            {
+                if (type.IsValueType && !(type.IsGenericType && type.GetGenericTypeDefinition() == typeof(Nullable<>)))
+                {
+                    return TryCatch<object>.FromException(Visitor.Exceptions.ExpectedReferenceOrNullableType);
+                }
```
Comment (neildsh) [inline Microsoft.Azure.Cosmos/src/Json/JsonSerializer.cs:309]: perhaps we can remove this

File: Microsoft.Azure.Cosmos/src/CosmosElements/CosmosElement.cs [hunk 4]
```diff
@@ -272,45 +272,49 @@ private sealed class CosmosElementToTypeOrder : ICosmosElementVisitor<int>
             private CosmosElementToTypeOrder()
             {
             }
+            public int Visit(CosmosUndefined cosmosUndefined)
+            {
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/CosmosElements/CosmosElement.cs:276]: newline

File: Microsoft.Azure.Cosmos/src/CosmosElements/CosmosElement.cs [hunk 5]
```diff
@@ -321,6 +325,10 @@ private sealed class CosmosElementWithinTypeComparer : ICosmosElementVisitor<Cos
             private CosmosElementWithinTypeComparer()
             {
             }
+            public int Visit(CosmosUndefined cosmosUndefined, CosmosElement input)
+            {
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/CosmosElements/CosmosElement.cs:329]: newline

**Comment by sboshra** [comment]
>                 {

we should either assert that we should never get an undefined, or we need to skip properties w/ undefined value

Samer for arrays

---
Refers to: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/Distinct/DistinctHash.cs:185 in 24feba8. [](commit_id = 24feba8aaa0a5785bc51936b92ad9a4897084201, deletion_comment = False)

File: Microsoft.Azure.Cosmos/src/CosmosElements/CosmosUndefined.cs [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+        public bool Equals(CosmosUndefined other)
+        {
+            return true;
+        }
+
+        public override int GetHashCode()
+        {
+            return Instance.GetHashCode();
+        }
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/CosmosElements/CosmosUndefined.cs:60]: consider 0

File: Microsoft.Azure.Cosmos/src/CosmosElements/CosmosUndefined.cs [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+        public override void WriteTo(IJsonWriter jsonWriter)
+        {
+        }
+
+        public static CosmosUndefined Create()
+        {
+            return Instance;
+        }
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/CosmosElements/CosmosUndefined.cs:69]: check where this function needs to be

**Review by sboshra — DISMISSED** [review]
:shipit:

File: Microsoft.Azure.Cosmos/src/CosmosElements/CosmosUndefined.cs (hunk 7 above)
Comment (neildsh) [reply Microsoft.Azure.Cosmos/src/CosmosElements/CosmosUndefined.cs:69]: There seems to be no clear pattern. However, `CosmosObject` does seem to put `static`s at the end

**Review by kirankumarkolli — APPROVED** [review]
Signing off on-behalf of Samer (as per offline conversation)

**Pull request merged**